dns/psl.py

```diff
@@ -0,0 +1,194 @@
+
+    The *base domain*, also known as the *registerable domain* or the *eTLD + 1* is one
+    level deeper than the public suffix.  For example, for `www.dnspython.org` the
+    public suffix is `org` and the base domain is `dnspython.org`.  Names which are
+    public suffixes do not have a base domain.
+
+    The *reduced domain* of a name is the base domain of that name if it is defined, or
+    the name itself otherwise.  Reduced domains are useful for statistical aggregations
+    where you are principally trying to aggregate by base domain but don't want to lose
+    track of queries without base names (e.g. queries to com, or the root).
+    """
+
+    def __init__(
+        self,
+        filename: str,
+        categories: Optional[Set[str]] = None,
+        allow_unlisted_gtlds: bool = True,
+        download_if_needed: bool = False,
+        url: str = PSL_URL,
+    ):
+        """Initialize a public suffix list.
+
+        *filename*, a ``str``, is the filename of the public suffix list, in the
+        `standard format <https://github.com/publicsuffix/list/wiki/Format>`. If the
+        file does not exist and *download_if_needed* has been specified, then the file
+        will be downloaded from the specified *url*.
+
+        *categories*, a set of ``str`` or ``None``, the PSL categories to include when
+        searching.  If ``None``, the default set ``{"ICANN", "PRIVATE"}`` is used.  The
+        ``"ICANN"`` category is the public suffixes administered by global and national
+        registries, and the "PRIVATE" category is public suffixes administred by private
+        entities as part of their namespace.
+
+        *allow_unlisted_gtlds*, a ``bool``, with a default of ``True``.  If ``True``,
+        then the root node is a wildcard node, and gTLDs not listed in the
+        public suffix database will still be considered as public suffixes.  For
+        example, a query of "www.example.bogus-gtld." would have a public suffix
+        of "bogus-gtld." and a base domain of "example.bogus-gtld.".  If ``False``,
+        then ``None`` will be returned for gTLDs which are not listed.
+
+        *download_if_needed*: a ``bool``, defaulting to ``True``.  If ``True``, then
+        download the list from the *url* if *filename* does not exist.  If ``False``,
+        then *filename* must exist.
+
+        *url*: a ``str``.  The URL to use if downloading the public suffix list is
+        required; the default is the standard URL recommended by publicsuffix.org.
+        """
+
+        if download_if_needed:
+            if not _have_httpx:
+                raise ValueError(
+                    "download_if_needed is True but httpx is not available"
+                )
+            if not os.path.isfile(filename):
+                response = httpx.request("GET", url)
+                if response.status_code == 200:
+                    with open(filename, "w") as f:
+                        f.write(response.text)
```

Comment:
Is there some reason this couldn't use `urllib.request()`, or is it doing something that requires `httpx`?